src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -191,12 +191,27 @@ public SourceGeneratedDocumentState WithGenerationDateTime(DateTime generationDa
 
     public SourceGeneratedDocumentState WithSyntaxRoot(SyntaxNode newRoot)
     {
-        // See if we can reuse this instance directly
-        if (this.TryGetSyntaxTree(out var tree) && tree == newRoot.SyntaxTree)
-            return this;
+        SyntaxTree newTree;
+        if (this.TryGetSyntaxTree(out var tree))
+        {
+            // See if we can reuse this instance directly
+            if (tree.TryGetRoot(out var root) && root == newRoot)
+                return this;
+
+            newTree = tree.WithRootAndOptions(newRoot, ParseOptions);
+        }
+        else
+        {
+            // We don't have a tree yet, so we need to create one. Using "newRoot.SyntaxTree" is the last resort though
+            // as it could lazily create one, which doesn't know about our parse options.
+            var factory = LanguageServices.GetService<ISyntaxTreeFactoryService>();
+            newTree = factory is not null
+                ? factory.CreateSyntaxTree(Attributes.SyntaxTreeFilePath, ParseOptions, SourceText, SourceText.Encoding, SourceText.ChecksumAlgorithm, newRoot)
```

Comment:
Isn't this SourceText the old SourceText? Does this need to take null?

Reply:
This is the source text we're going to pass to the Create method, so yes its the old one, but its also the new one. It's a quirk due to the `WithFrozenSourceGeneratedDocuments` always having to have a source text, because there is no truly lazy way to create a source generated document without a source text, so the caller just does `newRoot.GetText()`

Reply:
Okay, I pushed the `newRoot.GetText()` call into `SourceGeneratedDocumentState`, so this now looks correct (though I maintain it was _technically_ correct before too :P). It was always a bit of a smell having it be done externally anyway, so it is nicer now. Though was more changes than I was expecting :)